source/util/timer.cpp

```diff
@@ -0,0 +1,133 @@
+
+#include "util/timer.h"
+
+#include <sys/resource.h>
+#include <sys/time.h>
+#include <iomanip>
+#include <iostream>
+#include <map>
+#include <string>
+
+namespace spvutils {
+
+#if defined(SPIRV_TIMER_ENABLED)
+
+// A map data structure to search a CumulativeTimer object by its name. The key
+// of the map is the name and the value is the CumulativeTimer object.
+static std::map<const char*, CumulativeTimer* > CumulativeTimerMap;
+
+// Print the description of resource types measured by Timer class. If |out| is
+// NULL, it does nothing. Otherwise, it prints resource types. The second is
+// optional and if it is true, the function also prints resource type fields
+// related to memory. Its default is false. In usual, this must be placed before
+// calling Timer::Report() to inform what those fields printed by
+// Timer::Report() indicate.
+void PrintTimerDescription(std::ostream* out, bool measure_mem_usage) {
+  if (out) {
+    *out << std::setw(30) << "PASS name" << std::setw(12) << "CPU time"
+         << std::setw(12) << "WALL time" << std::setw(12) << "USR time"
+         << std::setw(12) << "SYS time";
+#if defined(SPIRV_MEMORY_MEASUREMENT_ENABLED)
+    if (measure_mem_usage) {
+      *out << std::setw(12) << "RSS" << std::setw(12) << "Pagefault";
+    }
+#endif  // defined(SPIRV_MEMORY_MEASUREMENT_ENABLED)
+    *out << std::endl;
+  }
+}
+
+// Do not change the order of invoking system calls. We want to make CPU/Wall
+// time correct as much as possible. Calling functions to get CPU/Wall time must
+// closely surround the target code of measuring.
+void Timer::Start() {
+  if (report_stream_) {
+    if (getrusage(RUSAGE_SELF, &usage_before_) == -1) {
+      usage_status_ = kClockGettimeFailed;
+    } else if (clock_gettime(CLOCK_MONOTONIC, &wall_before_) == -1) {
+      usage_status_ = kClockGettimeFailed;
+    } else if (clock_gettime(CLOCK_PROCESS_CPUTIME_ID, &cpu_before_) == -1) {
+      usage_status_ = kGetrusageFailed;
+    }
+  }
+}
+
+// The order of invoking system calls is important with the same reason as
+// Timer::Start().
+void Timer::Stop() {
+  if (report_stream_ && usage_status_ == kSucceeded) {
+    if (clock_gettime(CLOCK_PROCESS_CPUTIME_ID, &cpu_after_) == -1) {
+      usage_status_ = kClockGettimeFailed;
+    } else if (clock_gettime(CLOCK_MONOTONIC, &wall_after_) == -1) {
+      usage_status_ = kClockGettimeFailed;
+    } else if (getrusage(RUSAGE_SELF, &usage_after_) == -1) {
+      usage_status_ = kGetrusageFailed;
+    }
+  }
+}
+
+void Timer::Report(const char* tag) {
+  if (!report_stream_) return;
+
+  switch (usage_status_) {
+    case kGetrusageFailed:
+      *report_stream_ << std::setw(30) << tag
+                      << " ERROR: calling getrusage() fails";
+      return;
+    case kClockGettimeFailed:
+      *report_stream_ << std::setw(30) << tag
+                      << " ERROR: calling clock_gettime() fails";
+      return;
+    default:
+      break;
+  }
+
+  report_stream_->precision(2);
```

Comment:
We have much higher precision for cpu and wall times. Smaller optimization passes could take tens of microseconds, outputting 0.019ms is better than 0.00s.

I recommend output in milliseconds with three digits after decimal point for cpu and wall times, and integral milliseconds for user and system time.

Reply:
Done.

Reply:
Now it's outputting 0.000s.

I'm not strongly insisting on millisecond output, but it should be better than this:
```
                     PASS name    CPU time   WALL time    USR time    SYS time         RSS   Pagefault
inline-entry-points-exhaustive       0.001       0.001       0.000       0.000           0          13
   convert-local-access-chains       0.003       0.003       0.004       0.000         720          84
  eliminate-local-single-block       0.001       0.001       0.000       0.000           0           0
  eliminate-local-single-store       0.000       0.000       0.000       0.000           0           2
      eliminate-insert-extract       0.000       0.000       0.000       0.000           0           0
eliminate-dead-code-aggressive       0.004       0.004       0.004       0.000           0           0
       eliminate-dead-branches       0.000       0.000       0.000       0.000           0           0
                  merge-blocks       0.000       0.000       0.000       0.000           0           0
  eliminate-local-single-block       0.001       0.001       0.000       0.000           0           0
  eliminate-local-single-store       0.000       0.000       0.000       0.000           0           0
   eliminate-local-multi-store       0.000       0.000       0.000       0.000           0           2
      eliminate-insert-extract       0.000       0.000       0.000       0.000           0           0
eliminate-dead-code-aggressive       0.001       0.001       0.000       0.000           0           0
      eliminate-common-uniform       0.001       0.001       0.004       0.000           0           0
```

Maybe tweak the precision based on the measured CPU time? If measured 95 micros, output 0.000095. If measured 95 millis, output 0.095.

Also, note that if you are measuring a duration of 0.5ms, with 12.5% probability you would output:
CPU time: 0.000
WALL time: 0.000
USR time: 0.004

Reply:
How about using scientific floating-point notation instead (http://www.cplusplus.com/reference/ios/scientific/)?